File: client/src/partition/connection.rs
use tokio::net::TcpStream;

use wombatcore::{Header, ProduceRecord, Type};

pub struct Connection {
    socket: TcpStream,
}

// TODO(AD) Batch records.
impl Connection {
    pub async fn new() -> Connection {
        Connection {
            socket: TcpStream::connect("127.0.0.1:3110").await.unwrap(),
        }
    }

    pub async fn send(&mut self, record: ProduceRecord) {
        Header::new(Type::Produce)
            .write_to(&mut self.socket)
            .await
            .unwrap();
        record.write_to(&mut self.socket).await.unwrap();
    }
}

File: client/pkg/consumer/connection.go
package consumer

import (
	"net"

	"github.com/dunstall/wombat/client/pkg/record"
)

// TODO(AD) Set timeout on socket
type connection struct {
	sock net.Conn
}

func connect(addr string) (connection, error) {
	sock, err := net.Dial("tcp", addr)
	return connection{sock}, err
}

func (c *connection) send(request record.ConsumeRequest) error {
	_, err := c.sock.Write(request.Encode())
	return err
}

func (c *connection) receive() (record.ConsumeRecord, error) {
	return record.ReadConsumeRecord(c.sock)
}
